Factor damped wave gauge functions for reuse

nilsdeppe opened the pull request:
## Proposed changes

- factor out the weight and log factor functions from the DH file. This makes the file DH file a lot easier to navigate and will make it easier to reuse the functions in other damped wave-like gauges.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [x] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Factor damped wave gauge functions for reuse

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/GaugeSourceFunctions/DampedWaveHelpers.hpp:
```diff
@@ -0,0 +1,134 @@
+ * The spatial weight function is:
+ * \f{align*}{
+ *   W(x^i) = \exp(- (r / \sigma_r)^2)
+ * \f}
```
Please define r in terms of x^i

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/GaugeSourceFunctions/DampedWaveHelpers.hpp:
```diff
@@ -0,0 +1,134 @@
+
+/*
+ * Spacetime derivatives of the spatial weight function that enters the
+ * damped harmonic gauge source function.
```
Please emphasize that this assumes the time derivative is zero

geoffrey4444 commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/GaugeSourceFunctions/DampedHarmonic.py:
```diff
@@ -2,26 +2,16 @@
-    DW = np.zeros(len(coords) + 1)
-    DW[1:] = -2. * W / r_max / r_max * coords
-    return DW
+from .DampedWaveHelpers import (spatial_weight_function,
```
Why the . before DampedWaveHelpers here?

nilsdeppe replied:
That tells python to load the file relative to the current file's directory rather than check the python source paths. Going from the python source paths would make this line something like 120 chars long

geoffrey4444 reviewed (changes requested):
Thanks for doing this! Just some minor comments

moxcodes commented on src/Evolution/Systems/GeneralizedHarmonic/GaugeSourceFunctions/DampedWaveHelpers.cpp:
```diff
@@ -0,0 +1,328 @@
+    const Scalar<DataType>& lapse,
+    const Scalar<DataType>& sqrt_det_spatial_metric,
+    const double exponent) noexcept {
+  Scalar<DataType> logfac{};
```
Initialize with size to avoid forcing the `not_null` version into the `UNLIKELY`.

moxcodes commented on src/Evolution/Systems/GeneralizedHarmonic/GaugeSourceFunctions/DampedWaveHelpers.cpp:
```diff
@@ -0,0 +1,328 @@
+void spatial_weight_function(const gsl::not_null<Scalar<DataType>*> weight,
+                             const tnsr::I<DataType, SpatialDim, Frame>& coords,
+                             const double sigma_r) noexcept {
+  if (UNLIKELY(get_size(get(*weight)) != get_size(get<0>(coords)))) {
```
you could use `destructive_resize_components` (same everywhere below this pattern appears)

moxcodes commented on src/Evolution/Systems/GeneralizedHarmonic/GaugeSourceFunctions/DampedWaveHelpers.cpp:
```diff
@@ -0,0 +1,328 @@
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi,
+    const double exponent) noexcept {
+  tnsr::a<DataType, SpatialDim, Frame> d4_logfac{};
```
initialize with size to avoid the `UNLIKELY` branch (same below for this pattern everywhere)

moxcodes reviewed (dismissed):
Looks fine, just a few small things you could update while you're modifying these functions.

nilsdeppe committed:
Use destructive_resize in damped wave helpers

Also switch to allocating in the return-by-value function to avoid having to
take the UNLIKELY branch in the not_null functions.

nilsdeppe commented:
@moxcodes @geoffrey4444 thanks for the review comments!

I squashed in the documentation changes and added a new commit for the changes that use `destructive_resize` and allocate in the return-by-value functions. I figured it's easier if the one commit doesn't change the body of the functions in case we need to go back through the history.

I've also squashed in the fixes for clang-tidy complaints.

moxcodes reviewed (approved):
Looks good, thanks!
